cdap-watchdog/src/main/java/io/cdap/cdap/metrics/collect/AggregatedMetricsEmitter.java

```diff
@@ -33,38 +34,61 @@
   private final String name;
   // metric value
   private final AtomicLong value;
-  // specifies if the metric type is gauge or counter
-  private final AtomicBoolean gaugeUsed;
+
+  private volatile MetricType metricType = MetricType.COUNTER;
+  private Distribution distribution;
```

Comment:
It's a mix of Atomics, volatile and synchronized. Can we used just some one approach, preferably lock-free CAS-based.

Reply:
Atomics and volatile are used in the gauge and counter path. synchronized in the distribution metric path. For distributions, add and emit should be mutually exclusive. This can't be done using Atomics or volatile.

synchronized can be used everywhere, but I am worried it will slow down the gauge/increment which is 99% of the metrics. I am not sure what is the perf impact in practice. I can create a TODO/JIRA to test.

Reply:
You can still use AtomicReference to achieve that without sync block.

Reply:
I tried that earlier. If event(..) and emit() are called simultaneously, getMetricValue inside emit might finish before event (which calls add) is done. Effectively the add is lost.

Reply:
Basically:

```java
private final AtomicReference<Distribution> distributionRef = new AtomicReference<>(new Distribution());

public void event(long value) {
  Distribution distribution = distributionRef.get();
  ...
}

public MetricValue emit() {
  ...
  Distribution distribution = distributionRef.getAndSet(new Distribution());
  ...
}
```

Similarly for metricType, you can use `compareAndSet` in one of the `count`, `gauge`, and `event` method, and a `getAndSet` in the emit method.

Reply:
To elaborate
```
private final AtomicReference<Distribution> distributionRef = new AtomicReference<>(new Distribution());

public void event(long value) {
  Distribution distribution = distributionRef.get(); // Line 1
  distribution.add(...); // Line 2
  ...
}

public MetricValue emit() {
  ...
  Distribution distribution = distributionRef.getAndSet(new Distribution());
  distribution.getMetricValue();
  ...
}
```
Consider this race condition:
thread1 calls event(...). But before add is called in Line 2, thread2 calls emit(). Emit() finishes first. 
Then thread1 continues, but add(...)  logic is executed on an older value of distribution and is not reflected in any MetricValue.

Reply:
You can do a compareAndSet after adding in the `event` method to eliminate the race.

Reply:
We are not doing a getAndSet in `add`. Are you suggesting we do the getAndSet after `getMetricValue`? Even that will not help. After thread2 does getMetricValue and before getAndSet, thread1 might call add.

Reply:
You can do something like this to be sync-free:

```
public void event(long value) {
    Distribution distribution = distributionRef.getAndSet(null);
    while (distribution == null) {
      Thread.yield();
      distribution = distributionRef.getAndSet(null);
    }
    distribution.add(value);
    distributionRef.set(distribution);
  }


  @Override
  public MetricValue emit() {
    ...
    Distribution distribution = distributionRef.getAndSet(null);
    while (distribution == null) {
      Thread.yield();
      distribution = distributionRef.getAndSet(null);
    }
    MetricValue result = distribution.getMetricValue(name);
    distributionRef.set(new Distribution());
    return result;
  }
```

Reply:
I have never used Thread.yield() before, so I learnt it and implemented what you recommended. In fact, I wrote a concurrency test to compare the performance of both. Both implementations are left in the AggregatedMetricsEmitter class.

With synchronized, the concurrency test did 98 million adds in 10 seconds. Without synchronized, it did 284 million. The sync free code is definitely faster. But still, the cost of an add with synchronized is 0.1 microseconds, while the sync free add takes 0.033 microseconds. The performance gain is negligible.

I am worried that the Thread.yield() can lead to high cpu usage if for whatever reason one of add or emit take longer.  I found this stackoverflow article that recommends using Thread.yield in rare circumstances https://stackoverflow.com/questions/24948791/what-is-fast-wait-notify-or-busy-wait-in-java. 

I did a search as well in google3. Thread.yield usages are much lesser than synchronized. I found about 2K usages of thread.yield (many of them in test code) whereas synchronized was 98K.

Please let me know what you think. If you still feel we should use Thread.yield(), I will make the change.

Reply:
The first `new Distribution()` can be done in the constructor, which I did in my implementation.

Reply:
I think `emit()` could in theory wait a long time for value to come. While it is waiting, it will consume CPU.

Reply:
I agree. I think `synchronized` is safer. The `synchronized` version take 0.1 microseconds while the CAS/spinlock version takes 0.03 microseconds. The spinlock version is faster, but both have a negligible cost.

Reply:
Ideally, if we can use immutable to do it, that would be the lightest. The call to emit could be eventually 10s of thousand sec when a handler is handling that many requests per sec. We can keep the sync for now and evolve the code later (if we make the Distribution object immutable and using CAS inside, probably we can get away with any form of locking).

Reply:
Makes sense. Added a JIRA issue https://cdap.atlassian.net/browse/CDAP-18792 and created a TODO in the code with the JIRA issue.